🎉 New Source: Sumo Logic

iamzjk opened the pull request:
## What
This is a new source connector for [hacktoberfest-2021 issue #78](https://github.com/airbytehq/hacktoberfest-2021/issues/78)

## How
It uses the official [sumologic-python-sdk](https://github.com/SumoLogic/sumologic-python-sdk).
As a source connector of Sumo Logic [Search Job API](https://help.sumologic.com/APIs/Search-Job-API/About-the-Search-Job-API)


![Source setting page](https://user-images.githubusercontent.com/11445672/138494797-42d86d27-ebbc-40e7-bb28-23eb396447e1.png)
![Connection page](https://user-images.githubusercontent.com/11445672/138494809-796dfab6-c577-452d-b653-b454e05384fd.png)

## Recommended reading order

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated 
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/SUMMARY.md`
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [x] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

iamzjk commented:
```
Test session starts (platform: darwin, Python 3.7.11, pytest 6.2.5, pytest-sugar 0.9.4)
cachedir: .pytest_cache
rootdir: /Users/jzhuan/Forks/airbyte, configfile: pytest.ini
plugins: sugar-0.9.4, mock-3.6.1, timeout-1.4.2
collecting ... 
 airbyte-integrations/connectors/source-sumologic/unit_tests/test_client.py::test_check ✓                                                                                                                                                      8% ▉  
 airbyte-integrations/connectors/source-sumologic/unit_tests/test_client.py::test_search ✓                                                                                                                                                    17% █▋        
 airbyte-integrations/connectors/source-sumologic/unit_tests/test_incremental_streams.py::test_cursor_field ✓                                                                                                                                 25% ██▌       
 airbyte-integrations/connectors/source-sumologic/unit_tests/test_incremental_streams.py::test_cursor_field__when_by_receipt_time ✓                                                                                                           33% ███▍      
 airbyte-integrations/connectors/source-sumologic/unit_tests/test_incremental_streams.py::test_get_updated_state ✓                                                                                                                            42% ████▎     
 airbyte-integrations/connectors/source-sumologic/unit_tests/test_incremental_streams.py::test_stream_slices ✓                                                                                                                                50% █████     
 airbyte-integrations/connectors/source-sumologic/unit_tests/test_incremental_streams.py::test_supports_incremental ✓                                                                                                                         58% █████▉    
 airbyte-integrations/connectors/source-sumologic/unit_tests/test_incremental_streams.py::test_source_defined_cursor ✓                                                                                                                        67% ██████▋   
 airbyte-integrations/connectors/source-sumologic/unit_tests/test_incremental_streams.py::test_stream_checkpoint_interval ✓                                                                                                                   75% ███████▌  
 airbyte-integrations/connectors/source-sumologic/unit_tests/test_source.py::test_check_connection ✓                                                                                                                                          83% ████████▍ SDK Endpoint https://api.sumologic.com/api

 airbyte-integrations/connectors/source-sumologic/unit_tests/test_source.py::test_streams ✓                                                                                                                                                   92% █████████▎SDK Endpoint https://api.sumologic.com/api

 airbyte-integrations/connectors/source-sumologic/unit_tests/test_streams.py::test_init ✓                                                                                                                                                    100% ██████████


Results (0.71s):
      12 passed
```

iamzjk commented:
`python -m pytest integration_tests -p integration_tests.acceptance`

```
Test session starts (platform: darwin, Python 3.7.11, pytest 6.2.5, pytest-sugar 0.9.4)
cachedir: .pytest_cache
rootdir: /Users/jzhuan/Forks/airbyte, configfile: pytest.ini
plugins: sugar-0.9.4, mock-3.6.1, timeout-1.4.2
collecting ... 

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_match_expected[inputs0] ✓                                                                                                          7% ▊         
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_required[inputs0] ✓                                                                                                               14% █▌        
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_optional[inputs0] ✓                                                                                                               21% ██▎       
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_has_secret[inputs0] ✓                                                                                                             29% ██▉       
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_secret_never_in_the_output[inputs0] ✓                                                                                             36% ███▋    
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_oauth_flow_parameters[inputs0] ✓                                                                                                  43% ████▍    
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs0] ✓                                                                                                            50% █████    
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs1] ✓                                                                                                            57% █████▊   
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_discover[inputs0] ✓                                                                                                          64% ██████▌  
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_cursors_exist_in_schema[inputs0] ✓                                                                                   71% ███████▎ 
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_read[inputs0] ✓                                                                                                              79% ███████▉  
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_full_refresh.py::TestFullRefresh.test_sequential_reads[inputs0] ✓                                                                                        86% ████████▋
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_two_sequential_reads[inputs0] ✓                                                                                     93% █████████▍
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_state_with_abnormally_large_values[inputs0] ✓                                                                      100% ██████████

Results (53.40s):
      14 passed
```

iamzjk commented:
@marcosmarxm Please see integration tests output above ^ :)

marcosmarxm commented:
I'll take a look later today!

alafanechere commented on airbyte-integrations/connectors/source-sumologic/unit_tests/test_client.py:
```diff
@@ -0,0 +1,53 @@
+            mocker.call(search_job, limit=10000, offset=0),
+            mocker.call(search_job, limit=10000, offset=10000),
+        ]
+    )
```
please also test the others functions of your `Client`

iamzjk replied:
added tests for all other `Client` functions

alafanechere commented on docs/integrations/sources/sumologic.md:
```diff
@@ -0,0 +1,42 @@
+
+## Overview
+
+The Sumo Logic source supports both Full Refresh and Incremental syncs of their [Job Search API](https://help.sumologic.com/APIs/Search-Job-API/About-the-Search-Job-API).
```
```suggestion
The Sumo Logic source supports incremental syncs of their [Job Search API](https://help.sumologic.com/APIs/Search-Job-API/About-the-Search-Job-API).
```
It looks like you only implemented an incremental stream.

alafanechere reviewed (changes requested):
Hi @iamzjk thanks for your contribution. Please have a look at my first review comments. It'd be great to have a clear explanation on how you expect your incremental stream to behave in Airbyte, I'm a bit confused about your `to_time` parameter in an incremental context.

iamzjk commented on airbyte-integrations/connectors/source-sumologic/bootstrap.md:
```diff
@@ -0,0 +1,34 @@
+This connector uses sumologic python sdk https://github.com/SumoLogic/sumologic-python-sdk
+The sdk is a wrapper of sumologic API https://help.sumologic.com/APIs
+This connector implements the most common use case of sumologic: query and download. See seach job API details here: https://help.sumologic.com/APIs/Search-Job-API/About-the-Search-Job-API
```
Yes, I think renaming to `sumo-logic-search-jobs` probably makes more sense. Seach Job is a standalone API, also it is the one people often use to get their logs out of sumo-logic. other APIs are more for management.

iamzjk replied:
I see we have several Zendesk connectors, `source-zendesk-chat`, `source-zendesk-support`, and `source-zendesk-talk`. Sumo Logic has pretty much the same setup, so I gonna name this one `source-sumo-logic-search-job` to match the convention.

iamzjk commented on airbyte-integrations/connectors/source-sumologic/README.md:
```diff
@@ -0,0 +1,132 @@
+# Sumologic Source
```
Sure, will do

iamzjk replied:
As mentioned in another comment. I will rename this connector to `source-sumo-logic-search-job` if you agree. Will also update all `sumologic` to `sumo-logic` when I name the connector name change.

iamzjk commented on airbyte-integrations/connectors/source-sumologic/source_sumologic/spec.json:
```diff
@@ -0,0 +1,60 @@
+      "query": {
+        "type": "string",
+        "title": "Query",
+        "description": "Sumo Logic query string for the search job. You may need to escape certain characters."
```
Sure, I will add this query syntax document [link](https://help.sumologic.com/05Search/Get-Started-with-Search/How-to-Build-a-Search/Search-Syntax-Overview)

The query can take almost anything, a complex formatted query following the document above, or just a simple string match.

iamzjk commented on airbyte-integrations/connectors/source-sumologic/source_sumologic/spec.json:
```diff
@@ -0,0 +1,60 @@
+        "title": "To Time",
+        "pattern": "^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}$",
+        "description": "UTC date and time in the format 2017-01-25T00:00:00. Any data after this date will not be replicated. Leave empty if you want to pull util the most recent.",
+        "examples": ["2017-01-25T00:00:00"]
```
Makes sense. The reason I included `to_time` is because this param required by sumo-logic search job API https://help.sumologic.com/APIs/Search-Job-API/About-the-Search-Job-API#query-parameters

I think I can hide the logic, not allow users to set it, just always auto-generate a to_time in the code.

iamzjk replied:
Addressed. `to_time` is now handled by default, not an option anymore.

iamzjk commented on airbyte-integrations/connectors/source-sumologic/source_sumologic/source.py:
```diff
@@ -0,0 +1,94 @@
+        records = self.client.search(
+            query=self.config["query"],
+            from_time=from_time,
+            to_time=self.config.get("to_time", datetime.datetime.utcnow().replace(microsecond=0).isoformat()),
```
Implemented a new way to check the connection.

Now it checks the status of job id `0` (which is always an invalid job_id in sumo-logic)
It should return 404 if the user has correct permission, return 401 for bad API KEY.

iamzjk commented on airbyte-integrations/connectors/source-sumologic/unit_tests/test_client.py:
```diff
@@ -0,0 +1,53 @@
+def test_check(mocker):
+    mocker.patch("source_sumologic.client.SumoLogic")
+    Client("foo", "bar").check()
+
```
Improved this logic and tests. Implemented a new way to check the connection.

Now it checks the status of job id 0 (which is always an invalid job_id in sumo-logic)
It should return 404 if the user has correct permission, return 401 for bad API KEY.

iamzjk commented on airbyte-integrations/connectors/source-sumologic/source_sumologic/source.py:
```diff
@@ -0,0 +1,94 @@
+
+        try:
+            client = Client(config["access_id"], config["access_key"])
+            client.check()
```
Updated. When there is no proper permission, It raises the same exception from API now

iamzjk commented on airbyte-integrations/connectors/source-sumologic/source_sumologic/client.py:
```diff
@@ -0,0 +1,80 @@
+
+    def check(self):
+        # Make sure the client has correct access info
+        self.sumo.get_personal_folder()
```
A new logic has been implemented here to better check the connction

iamzjk commented on airbyte-integrations/connectors/source-sumologic/source_sumologic/spec.json:
```diff
@@ -0,0 +1,60 @@
+        "title": "Offset",
+        "description": "Skip the first N messages. Default 0",
+        "default": 0
+      }
```
updated. `offset` is no longer configurable by users.

iamzjk commented on airbyte-integrations/connectors/source-sumologic/source_sumologic/client.py:
```diff
@@ -0,0 +1,80 @@
+
+        delay = 5
+        status: dict = self.sumo.search_job_status(search_job)
+        while status["state"] != "DONE GATHERING RESULTS":
+            if status["state"] == "CANCELLED":
+                self.logger.warning("Search job status: CANCELLED.")
+                break
+            time.sleep(delay)
+            status = self.sumo.search_job_status(search_job)
```
Added doc for possible values of state field.
Added timeout strategy, default 8 hours to match sumo-logic API https://help.sumologic.com/APIs/Search-Job-API/About-the-Search-Job-API#session-timeout

iamzjk commented on airbyte-integrations/connectors/source-sumologic/source_sumologic/client.py:
```diff
@@ -0,0 +1,80 @@
+        # pagination logic
+        self.logger.info(f"Start downloading {total} messages...")
+        while offset < total:
+            response = self.sumo.search_job_messages(search_job, limit=limit, offset=offset)
```
Sumo Logic has 4 calls per second limit. Added logic to sleep between calls.

iamzjk commented on airbyte-integrations/connectors/source-sumologic/source_sumologic/source.py:
```diff
@@ -0,0 +1,94 @@
+        self,
+        sync_mode: SyncMode,
+        cursor_field: Optional[List[str]] = None,
+        stream_slice: Optional[Mapping[str, Any]] = None,
```
I like this idea, tho I haven't encountered any issue with my daily usage so far (I've been using this connector for my work). What do you think if we make this a future improvement?

alafanechere replied:
What you defined in this function, by reading only one minute of data, is already a kind of stream slicing. I'm afraid that if you're not defining `stream_slices` (of one minute interval for example) you might only read one minute data on each sync and might not retrieve all the messages.

iamzjk commented:
Hi @alafanechere, I addressed all the suggestions I made (hope so, let me know if I missed any one of them).

Regarding the naming. Please see this comment https://github.com/airbytehq/airbyte/pull/7283#discussion_r746886960
Let me know what you think. I will update it once you agree on the naming.

alafanechere commented:
> Hi @alafanechere, I addressed all of your comments (hope so, let me know if I missed any one of them).
> 
> Regarding the naming. Please see this comment [#7283 (comment)](https://github.com/airbytehq/airbyte/pull/7283#discussion_r746886960) Let me know what you think. I will update it once you agree on the naming.


The suggested naming looks good to me!
Thanks for your new commits 👍 , I'll go for a second review then.

alafanechere commented on airbyte-integrations/connectors/source-sumologic/acceptance-test-config.yml:
```diff
@@ -0,0 +1,24 @@
+    - config_path: "secrets/config.json"
+      configured_catalog_path: "integration_tests/configured_catalog.json"
+      future_state_path: "integration_tests/abnormal_state.json"
+  full_refresh:
+    - config_path: "secrets/config.json"
+      configured_catalog_path: "integration_tests/configured_catalog.json"
```
It appears that you have only incremental streams according to your streams declaration. Could you also update the `supported_sync_modes` accordingly in your integration tests catalog and configured catalog?
```suggestion
```

alafanechere commented on airbyte-integrations/connectors/source-sumologic/source_sumologic/source.py:
```diff
@@ -0,0 +1,106 @@
+class IncrementalSumologicStream(SumologicStream, ABC):
+
+    # Not confident with the order returned by sumologic, so don't check in until all data are read
+    state_checkpoint_interval = None
```
This not recommended to not checkpoint your stream in an incremental mode, otherwise use Full refresh streams... I suggest that you implement `stream_slices` to handle a smart checkpointing if you're not confident with the returned order. Doc [here](https://docs.airbyte.io/connector-development/cdk-python/stream-slices).

alafanechere commented on airbyte-integrations/connectors/source-sumologic/source_sumologic/client.py:
```diff
@@ -0,0 +1,112 @@
+            time.sleep(delay)
+            waittime += delay
+            if waittime > timeout:
+                raise Exception(f"Reached query timeout {waittime}/{timeout}.")
```
```suggestion
                raise Exception(f"Reached query timeout: waited for {waittime} seconds, timeout is {timeout} seconds.")
```

alafanechere commented on airbyte-integrations/connectors/source-sumologic/source_sumologic/client.py:
```diff
@@ -0,0 +1,112 @@
+        #   - CANCELLED
+        # https://help.sumologic.com/APIs/Search-Job-API/About-the-Search-Job-API#sample-session-1
+        while status["state"] != "DONE GATHERING RESULTS":
+            if status["state"] == "CANCELLED":
```
I think you could also add a check for the `FORCE PAUSED` status.

alafanechere commented on airbyte-integrations/connectors/source-sumologic/source_sumologic/client.py:
```diff
@@ -0,0 +1,112 @@
+        #   - CANCELLED
+        # https://help.sumologic.com/APIs/Search-Job-API/About-the-Search-Job-API#sample-session-1
+        while status["state"] != "DONE GATHERING RESULTS":
+            if status["state"] == "CANCELLED":
```
```suggestion
            if status["state"] in ["CANCELLED", "FORCE PAUSED"]:
```

alafanechere replied:
Why aren't you raising an exception here? I'm not sure that its an expected behavior...

alafanechere reviewed (changes requested):
I made a second review, mostly on the core. I'll focus on tests and other files once you renamed the connector as you suggested + implement / challenge my suggestions.

iamzjk commented:
I see today is the last day of the hacktoberfest. Looks like it is very unlikely it will be merged before the deadline. So kindly close this PR, I may release it later as a standalone connector, in case it is helpful for any other user.

marcosmarxm commented:
@iamzjk it's not a problem, all contributions made before 15th-nov will be eligible to receive the award. Thanks so much for your contribution 🚀

alafanechere commented:
HI @iamzjk do you still plan to work on this?

marcosmarxm commented:
I sent a message to the user: doesn't time at the moment but want to finish the connector. I'll add the label `blocked` @iamzjk ping me or Augustin when you want to continue this work.

marcosmarxm commented:
@iamzjk  I'll close this due to inactivity. If you want to return this work you can reopen the PR.
